app/src/main/java/org/olpc/xorobothash/MainActivity.java

```diff
@@ -0,0 +1,164 @@
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+        robotImageView = (ImageView) findViewById(R.id.robot_hash_image);
+        robotImagePath = this.getFilesDir() + File.pathSeparator + getString(R.string.robot_image_internal_filename);
+        getRobot();
+    }
+
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        // Inflate the menu; this adds items to the action bar if it is present.
+        getMenuInflater().inflate(R.menu.menu_main, menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        // Handle action bar item clicks here. The action bar will
+        // automatically handle clicks on the Home/Up button, so long
+        // as you specify a parent activity in AndroidManifest.xml.
+        int id = item.getItemId();
+
+        //noinspection SimplifiableIfStatement
+        if (id == R.id.action_settings) {
+            return true;
+        }
+
+        return super.onOptionsItemSelected(item);
+    }
+
+
+    private void getRobot() {
+        Log.i (tag, "In getRobot()");
+        //Check if the robot image file already exists
+        File file = new File(robotImagePath);
+        // Check if we have already downloaded the XO's robot image
+        if(file.exists()) {
+            Log.i (tag, "Robohash image file exists.");
+            // Show the already downloaded robohash image
+            showRobot(robotImagePath);
+            return;
+        }
+
+        Log.i (tag, "File does not exist.");
+        // We haven't already downloaded the image - check network connectivity
+        ConnectivityManager connMgr = (ConnectivityManager)
+                getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
+        if (networkInfo != null && networkInfo.isConnected()) {
+            Log.i (tag, "Network is connected.");
+            // Get the device's serial and concatenate into a RoboHash URL
+            String serialNumber = android.os.Build.SERIAL;  // if SERIAL doesn't work, try Settings.Secure.ANDROID_ID; ?
+            String roboHashUrl = getString(R.string.robohash_url) + serialNumber + getString(R.string.robohash_file_extension);
+            // Execute the async download of the image
+            new DownloadRobotImage().execute(roboHashUrl);
+        } else {
+            Log.e (tag, "Network is NOT connected.");
+            // Hide the "Your robot is" message
+            TextView robotText = (TextView) findViewById(R.id.robot_text_message);
+            robotText.setVisibility(View.GONE);
+            // Show the no network connection message
+            TextView errorMessage = (TextView) findViewById(R.id.error_message);
+            errorMessage.setText(getString(R.string.no_network_connection));
+            return;
+        }
```

Comment:
Following from the comment at 109, this whole method would simplify to something like:
```
boolean isConnected = ...;
boolean done = updateView(isConnected);
if (!done) {
  ...
  if (isConnected) {
     downloadRobot(); // Will call updateView again when done
  }
  // Nothing else to do, updateView already showed the right error message
}
```